File: 프로그래머스/2/178870. 연속된 부분 수열의 합/연속된 부분 수열의 합.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

vector<int> solution(vector<int> sequence, int k) {
    vector<int> answer;
    
    int left = 0, right = 0, sum = 0;
    int minLength = sequence.size() + 1, minLeft = 0, minRight = 0;
    
    while (right < sequence.size()) {

        while (right < sequence.size() && sum < k) {
            sum += sequence[right++];
        }
        
        while (sum > k && left < right) {
            sum -= sequence[left++];
        }
        
        if (sum == k) {
            int length = right - left;
            if (length < minLength) {
                minLength = length;
                minLeft = left;
                minRight = right-1; 
            }
        }
  
        sum -= sequence[left++];
    }
    
    answer.push_back(minLeft);
    answer.push_back(minRight);
    
    return answer;
}

File: 프로그래머스/2/178870. 연속된 부분 수열의 합/연속된 부분 수열의 합.py
def solution(sequence, k):
    answer = []
#     sum=0 
#     indexes = []
#     for i in range(0, len(sequence)):
#         for j in range(i, len(sequence)):
#             indexes.append((i,j)) 
    
#     _min = 1000000000000
#     for start, end in indexes:
#         sum=0 
#         for i in range(start, end+1):
#             sum = sum + sequence[i]
#         if sum == k:
#             if end-start < _min:
#                 _min = end-start
#                 answer = (start, end)
    
    n = len(sequence)
    left = 0
    curr_sum = 0
    best_len = float('inf')
    answer = [0, 0]

    for right in range(n):
        curr_sum += sequence[right] 

        while curr_sum > k and left <= right:
            curr_sum -= sequence[left]
            left += 1

        if curr_sum == k:
            if right - left < best_len:
                best_len = right - left
                answer = [left, right]
            # ※ 같은 길이의 구간이 나와도 갱신하지 않으므로
            #    최초(=시작 인덱스가 가장 작은) 구간이 유지됩니다.

    return answer